rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
+        let drv_data = Box::into_raw(drv_data) as *mut c_types::c_void;
+        Ok(drv_data) as KernelResult<_>
+    };
+    // TODO don't we need to pin this?
+    let ptr = match f() {
+        Ok(ptr) => ptr,
+        Err(e) => return e.to_kernel_errno(),
+    };
+    rust_helper_platform_set_drvdata(pdev, ptr);
+    0
+}
+
+fn new_of_device_id(compatible: &CStr<'static>) -> KernelResult<bindings::of_device_id> {
+    // TODO:
+    // - fail at build time if compatible CStr doesn't fit.
+    // - can we do this safely without transmute?
+    let mut buf = [0_u8; 128];
+    if compatible.len() > buf.len() {
+        return Err(Error::EINVAL);
+    }
+    // PANIC: this will never panic: `compatible` is not longer than `buf`.
+    buf[..compatible.len()].copy_from_slice(compatible.as_bytes());
+    Ok(bindings::of_device_id {
+        // SAFETY: re-interpretation from [u8] to [i8] of same length is always safe.
+        compatible: unsafe { transmute::<[u8; 128], [i8; 128]>(buf) },
+        ..Default::default()
+    })
+}
+
+unsafe extern "C" fn remove_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let ptr = rust_helper_platform_get_drvdata(pdev);
+    let drv_data: Box<T::DrvData> = Box::from_raw(ptr as _);
+    drop(drv_data);
+    0
+}
+
+/// A registration of a platform driver.
+#[derive(Default)]
+pub struct Registration {
+    registered: bool,
+    pdrv: bindings::platform_driver,
+    of_table: [bindings::of_device_id; 2],
+    _pin: PhantomPinned,
+}
+
+impl Registration {
+    fn register<P: PlatformDriver>(
+        self: Pin<&mut Self>,
+        name: CStr<'static>,
+        of_id: CStr<'static>,
+        module: &'static crate::ThisModule,
+    ) -> KernelResult {
+        // SAFETY: We must ensure that we never move out of `this`.
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            // Already registered.
+            return Err(Error::EINVAL);
+        }
+        // TODO should create a variable size table here.
+        this.of_table[0] = new_of_device_id(&of_id)?;
+        // SAFETY: `name` pointer has static lifetime.
+        // `of_table` points to memory in `this`, which lives as least as
+        // long as the `platform_device` registration.
+        // `module.0` lives as least as long as the module.
+        this.pdrv.driver.name = name.as_ptr() as *const c_types::c_char;
+        this.pdrv.driver.of_match_table = this.of_table.as_ptr();
+        this.pdrv.probe = Some(probe_callback::<P>);
+        this.pdrv.remove = Some(remove_callback::<P>);
+        let ret = unsafe { bindings::__platform_driver_register(&mut this.pdrv, module.0) };
```

Comment:
There was some discussion of the benefits of having vtables be `const` (e.g. (here)[https://github.com/Rust-for-Linux/linux/pull/228#issuecomment-828403944]. Maybe we'd also want to have `this.pdrv` point to a `const` vtable? We'd have the same issue as `file_operations` with the owner field though.

Reply:
Thanks Adam for linking to that fascinating discussion! I suspect in this case it may not work, as `__platform_driver_register()` wants `this.pdrv` to be mutable, not const. That function immediately modifies `this.pdrv`:

https://github.com/Rust-for-Linux/linux/blob/99bec9d3476d8cfdafda8ddf83cb8f985e5ea22b/drivers/base/platform.c#L884-L897

Reply:
Thanks, makes sense!